Theory assessment/Theory_assessment_Software_Engineering.md

```diff
@@ -0,0 +1,150 @@
+- BIGGEST/COMMON PROBLEMS:
+    - Confusion with names or people, if there is a lot of employees.
+    - People wanting to change their votes.
+- COMPONENTS/TOOLS:
+    - SQL for holding data and e-mails - who have voted.
+    - PYTHON for back-end logic.
+    - HTML and CSS for front-end.
+    - API for e-mail notifications.
+
+### 5. What's the software development cycle?
+A: Software development cycle known as SDLC, is needed to deliver expected goals to client. It has 6 steps:
+1. Planning: making plan(s) how to achieve the goal and requirements analysis.
+2. Analysis: deciding tool/methods that project is made and how these fill requirements.
+3. Design: planning and setting up the architecture, so design would have a suitable logic.
+4. Development: coding and making the goal product.
+5. Integration and testing: putting all the work pieces together and testing + documentation(architecture, technical, user).
+6. Maintenance: deploying, improving and maintaining the product.
+
+### 6. What's the difference between agile and waterfall?
+A: Waterfall is more phase-to-phase (linear) and strict approach to development. Waterfall is great for projects that have specific timeline and very concrete requirements, since everything is organized before work starts. Everything is made simple and project can be done without any hurdles. Cons of waterfall are that it is not very flexible, because if changes are made or something goes wrong, there is a chance to have to start again. It also takes time and cost can be big, if there is something that needs to be done fast. Agile is more flexible, there are not many phases and requires more testing, which leads to better customer satisfaction. Projects are divided into smaller pieces and prioritized. Cons of Agile are that there should be clear picture of the goal, team commitment and tight cooperation for project to succeed.
+ ![image](agilevswaterfall.jpg)
+
+### 7. What is a reduced function used for?
+A: Reduce function takes list and function of two arguments.
+Given functions is cumulatively applied to all list elements. 
+Reduce returns cumulated result(single value).
+```
+import functools 
+
+list= ['Code', 'First', 'Girls']
+full_name= functools.reduce(lambda x,y: x+y,list)
+print(full_name)
+
+Result: CodeFirstGirls
+```
+### 8. How does merge sort work?
+A: Merge sort is algorithm that takes array and divides it smaller and smaller sub-pieces until the dividing is impossible. Then it sorts the smallest pieces, compares them, sorts them and merges them into a list.   
+
+![image](merge_sort.jpg)
+
+### 9. Generators - Generator functions allow you to declare a function that behaves like an iterator, i.e. it can be used in a for loop. What is the use case?
+A: All generators are iterators. Generator is used for loops, so the loop would not be influenced by loop's iterations and returning values, so it is used for iterating through generator object and returns sequence of values.
```

Comment:
6/10 The explanation is correct, but the main idea of the question is the 'use case' of the generator. In which situation you would use a generator and why.